### ISSUE TITLE
pkcs7+der: mixed BER/DER encoding with indefinite length

### ISSUE THREAD

**Issue by smndtrl** [RustCrypto/formats#779]
Hi,

while looking into the `pkcs7` crate for a CMS usecase around the Apple world I discovered that their detached signatures use the BER indefinite length encoding for some of the `SEQUENCE` which of course is not supported by the `der` crate.

Has there been any thoughts/discussions around if that's something the formats repo should address or if that is out of scope (for now) and left to other crates.

I started with adding the signed-data content to `pkcs7` validating against a DER encoded signature and discovered the problem afterwards with the ones Apple generates :( [Link to example](https://lapo.it/asn1js/#MIAGCSqGSIb3DQEHAqCAMIACAQExCzAJBgUrDgMCGgUAMIAGCSqGSIb3DQEHAQAAoIIESTCCBEUwggItoAMCAQICARAwDQYJKoZIhvcNAQELBQAwTDELMAkGA1UEBhMCVVMxEDAOBgNVBAoTB3NjZXAtY2ExEDAOBgNVBAsTB1NDRVAgQ0ExGTAXBgNVBAMTEE1JQ1JPTURNIFNDRVAgQ0EwHhcNMjIxMTI5MTU0NDI4WhcNMjMxMTI5MTU0NDI4WjAXMRUwEwYDVQQDEwxhc2RmMTIzNGFzZGYwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQC8wlP2vnVUluoVmAibRcyrYzvcM42bXncKhrcin_NoXUI0nuOjUCeZtSkpXD9F6lP7dN1MzlAPI2OZBQ46hpGqxvFnLOBqE77Obq93D6JUssmE5VrVWkH_oPBKovV3Vj2QSqTUoZjSKI0l92qE3pMjJWfiXsz8Jvyd5s0iZJEiTWc89w5QME_tHqdjiCwWWnlTqjbE-aXojgD1CtmdCYGcRHUUwTs--1nfcbWgE9Lhc3u0MHlvDzGpiUj2fbNWY9T5rsQ5K4XK5rrUk9333zwW0QWWyBcGdFnCs2nCmR1MGQOBXaIGcCHgpTlfhyRKKzCl8btGFNDVJ6vA643CAKy9AgMBAAGjZzBlMA4GA1UdDwEB_wQEAwIHgDATBgNVHSUEDDAKBggrBgEFBQcDAjAdBgNVHQ4EFgQUogd1xrgbmo68wL6Zek-RhNUBdnMwHwYDVR0jBBgwFoAUBgrXSGTJePiZb5VBx4VslWnSFi0wDQYJKoZIhvcNAQELBQADggIBAFBlvlZXePm_Bx392Rgp1iwouG8freNTzWsV-YxrQQxK8NXAzz9P8XL5d-7rlAxvuTjyNiv0obOPpZP9rO_KettcFK3iy6t9MaQFFE82159lkuc_PQ3Eq4Pb_NqHuWCSOXF0XkbdIxiHBBzuxxQcvdETM8AWCQcNh7cEODvK5HqEEv5k2ztMurGZJxlVkcie60SmZE-jSSQx70n-EtdEj1751ldT_tnZSvcV5nw2bjYXnJ43IftkJJxj8WzpjLA9a0ClXERryWhPUvm3sdBjA3vvcl1r0cGQojCNwa_SSLj67Xg__XFvSpyEEiY7wZjAqHl5qOmKXaz0GJ5jlaBDrNp1TcP4KzzYz7kSvdVTz0ekn20j1U2uO4o98ocrHyISqYhj5bBL9izvSu5ff_tVkGnQ7NG3AE8wC6orc6IyXdLj8V2_1erT9r2r-feEFRwF9BXluAN8Ico1viK_tdryA9Zj7Px49DKmweCZJhFRwexfNjWQmuKmvA7L86--ttTurP-SYU9T6b1Hp7c6wjJvM9d2bkAn97xJTPHOu21f3ZckevpHd2C-lJ06WhKI_LAnGDYlFNfnBG6vjoaXp1G2-eTqgNG9BZZtAlIT-6gh5UMA6Iynf8a6-KgqZn84oFKYlYxxJZvoJZuk4U341wn7t_NPBHTaquLnJI7WB6VUJd3UMYIB1zCCAdMCAQEwUTBMMQswCQYDVQQGEwJVUzEQMA4GA1UEChMHc2NlcC1jYTEQMA4GA1UECxMHU0NFUCBDQTEZMBcGA1UEAxMQTUlDUk9NRE0gU0NFUCBDQQIBEDAJBgUrDgMCGgUAoF0wGAYJKoZIhvcNAQkDMQsGCSqGSIb3DQEHATAcBgkqhkiG9w0BCQUxDxcNMjIxMTI5MTU0NDMxWjAjBgkqhkiG9w0BCQQxFgQUkgRYlp6zAQLYAWVak1087XihS8cwDQYJKoZIhvcNAQEFBQAEggEAPxiqiDDtyXlk9y4M0GfIun8_QP1qTBgRnt0ovt4atjwDAFDc7anAMuxO18Z3IR6m-Eo8cQJigubgo2OYR6gR11pfbZajZNqfv5aUAwi8tjxIIo1djvUtJYGrc5JL88NiREasfJfssw6zNZ8RWYxY0w6SL99OGHvudpjmRSNrbknTzxDbu4bVmjPl70Cj1d_OEjHQoNR3kGdx0xl8NTsDm0YlvAbzoPV57lUEshq4cRWvWmNUtQKsxFgmcPnPkyEU4QrJpnSUKRQD-9MyPfjpru72kFZvXYLtHHKC_41Di9cnj1U9DWewR9TIrqHxwapEnME_MVfsl1IaQwBTewXz0QAAAAAAAA)

Simon

**Comment by tarcieri** [comment]
We've talked about introducing a lax mode to the `der` crate which supports a limited number of BER productions necessary to interop with real-world use cases like this. So far there hasn't been one until now.

**Comment by smndtrl** [comment]
I gave it a shot with my limited experience with Rust and also this codebase. It is meant as a starting point.

I used it successfully with changes to `pkcs7`/CMS crate that are not yet ready/

**Comment by tarcieri** [comment]
@smndtrl rather than trying to change how the existing parsing works, you can write your own "lax" implementation of `Decode` for a specific type

**Comment by smndtrl** [comment]
@tarcieri Wouldn't that require me changing how `DecodeValue` is implemented on `Sequence`? My PKCS7 data starts with `30 80` and that already makes the `Length::decode` fail without reaching the first `DecodeValue` code for `ContentInfo` in `pkcs7`.

For making indefinite length work a solution for `Length` and `Reader`s is required. The ASN1 I encountered had indefinite lengths for `SEQUENCE` and `OCTET STRING` but also can be other fields that can be constructed in BER.  For `OctetStringRef` I'm not sure how to handle those cases as it then is non-contiguous memory (or at least broken up with Tags/Lengths values. For `Sequence` it via the `Reader`.

I'm fine with carrying this patch on my own as it works well for my use case. I might revisit the suggested approach later on.

**Comment by tarcieri** [comment]
@smndtrl you'll need to write all the impls by hand if you go that route, and you will have to avoid impl'ing one of `DecodeValue` or `FixedTag` (I would suggest `DecodeValue` unless it needs to be in an IMPLICIT and context-specific), as otherwise it will receive a blanket impl of `Decode`.

If you can't figure it out I can take a look at some point based on your `lax` branch. It should be possible to impl the "lax" logic for just a single type.

**Comment by bkstein** [comment]
> We've talked about introducing a lax mode to the `der` crate which supports a limited number of BER productions necessary to interop with real-world use cases like this. So far there hasn't been one until now.

We've found a real-world use case: [`EJBCA`](https://www.ejbca.org) sends (degenerate certificates-only) CMS messages encoded with BER. Unfortunately, that's in line with RFC 5652:

```
   ...
   The CMS values are generated using ASN.1 [X.208-88], using BER-
   encoding (Basic Encoding Rules) [X.209-88].
```

**Comment by carl-wallace** [comment]
BER was an issue working on PKCS #12 as well. Exports from Firefox are BER. In a cert collection I use for testing, there are several several AIAs hosting P7s that are BER encoded.

**Comment by tarcieri** [comment]
We're aware this is a real issue. It's something I've been meaning to address.

**Comment by tarcieri** [comment]
I think we can probably use an approach like https://github.com/RustCrypto/formats/pull/810 but we need some way to control the granularity so enabling it doesn't cause `der` to revert to BER parsing globally

**Comment by bkstein** [comment]
FYI, I have implemented a simple parser/converter (`Berder`), that replaces all indefinite length specifications in a BER message by definite lengths. The parser recurses through the complete message (though there is a limit for recursion depth 🙂). It doesn't fix ordering, but as RustCrypto's `der` crate fixes it, I think it should work reliably. I tested it against EJBCA messages. The usage is simple. Just prepare your BER data before you try to parse it, e.g.:
```rust
    ...
    let der = Berder::ber_to_der(ber.as_slice())?; // Indefinite -> definite lengths
    let ci = ContentInfo::from_der(der.as_slice())?;
    let pki_message = SignedData::from_der(ci.content.to_der()?;
```
This could be implemented as `Decode::from_ber()`, but could also be integrated into `Decode::from_der()`. However, the latter would introduce some performance loss.
What do you think?

**Comment by tarcieri** [comment]
I think we can probably implement `Decode::from_ber` in a way that doesn't require transcoding the entire document to DER first

**Comment by bkstein** [comment]
> I think we can probably implement `Decode::from_ber` in a way that doesn't require transcoding the entire document to DER first, using an approach similar to #810 (we have since added an `IndefiniteLength` type), and perhaps carrying the flag for using BER-like rules on the `Reader`, for lack of a better option.

@tarcieri I have started doing that based on #810. I think I got pretty far with this implementation, but there are one or two issues left (correct handling of the end-of-content markers). The main idea was, to modify the `der` parser to do look-ahead evaluation of indefinite lengths. Whenever the parser encounters an indefinite length, it stores the current position and makes a "fast-forward" until the end of the current TLV is reached. Then, the length is "definite"ly known and we can resume parsing with the definite length at the stored position. This look-ahead must be recursive, even if e.g. an `OctetString` is parsed, because it might contain other (nested) TLVs with indefinite lengths.
I have uploaded this [to my repo](https://github.com/bkstein/RustCrypto-formats/tree/der/ber), if you want to have a look at it.
And as you mentioned `IndefiniteLength`: I changed `der::Header`:
```
pub struct Header {
    /// Tag representing the type of the encoded value
    pub tag: Tag,

    /// Length of the encoded value
    pub length: IndefiniteLength,
}
```
This implementation is not complete, but I think it is already more complete, than the current #810.

**Comment by tarcieri** [comment]
@bkstein I'd like to avoid making changes to the existing APIs like that. The idea would be to use `IndefiniteLength` as an encoding type to parse lengths in "BER mode", but then convert back to a `Length` in the public API.

However, I'd definitely like to also preserve the way the existing implementation works so indefinite lengths are rejected when parsing DER.

**Comment by bkstein** [comment]
Yes, the `Header` change would break compatibility. I saw, that signature's crates `dsa` and `ecdsa` access `Header::length`, so this would actually affect compatibility here.
However, I introduced a [flag `is_parsing_ber` in `Reader`](https://github.com/bkstein/RustCrypto-formats/blob/18bd1e9bfe9ecbe24853e4812ef8adef72f23e47/der/src/reader/slice.rs#L19). Is that approximately, what you intend with "BER mode"? In "DER mode", parsing is unchanged.

**Comment by tarcieri** [comment]
I was thinking an enum like `Encoding::Ber` which would improve clarity, I think, but yes that's the basic idea

**Comment by tarcieri** [comment]
I've bumped `der`'s version to v0.8.0-pre.0 so we're now allowed to make breaking changes, which should make this easier.

I opened #1321 which adds an `EncodingRules` enum with `Ber` and `Der` variants, as well as a `Reader::encoding_rules` method to interrogate them.

**Comment by Firstyear** [comment]
@tarcieri For reference, we were interested in the BER capability as we want to unify our LDAP and Kerberos parsers on one trusted encoding/decoding stack. Currently we have a "ber only" parser, but we would like to move to this crate. I'll do some testing soon to see what we need for extending this to support BER in our cases.

**Comment by dwhjames** [comment]
@tarcieri,

I hit this issue.

I did see
- https://github.com/RustCrypto/formats/pull/1321

And I read through prior attempts of @smndtrl and @bkstein 
- https://github.com/RustCrypto/formats/pull/810
- https://github.com/RustCrypto/formats/compare/master...bkstein:RustCrypto-formats:der/ber

I spent some time searching for additive-only changes, and even with, PR #1321, I could not see a path.

My impression from the codebase was that there would need to be breaking changes, and one possible path might look like,

1. using `IndefiniteLength` rather than `Length` within `Header`,
2. passing `Header` rather than `Length`, to function `read_nested` of trait `Reader`,

so that the existing API and usage doesn’t change dramatically, but so that nested readers can handle the constructed, indefinite-length method.

---

As I need a solution in the short term, I had to go with the ‘transcoding’ approach that was discussed above.

The following PR is me contributing that code back, in case the maintainers are open to accepting this into the crate in the near term (at least until more general solutions are developed). Or failing that, maybe it will be of use to others that find this issue while it remains open.
- https://github.com/RustCrypto/formats/pull/1652

**Comment by deknowny** [comment]
Any plans to support BER? Apple MDM is real pain in rust, there are no other ways really to implement a high-level solution to verify over-the-air device response

**Comment by tarcieri** [comment]
@deknowny yes, the goal is to support decoding a limited subset of BER before the next release, though to do that we need to change `Header` to have `IndefiniteLength` as its `length` field. See also #1321.

#810 is some prior art on what that would look like, if someone wants to attempt a PR.

**Comment by tarcieri** [comment]
I attempted what I hoped might be an interesting solution: adding a `BerReader` which handles indefinite lengths and otherwise presents the `Reader` interface.

Unfortunately `Reader::read_nested` uses an `FnOnce` that accepts `&mut Self` as an argument. While it's possible to cast to `&mut BerReader` from an inner generic type (using `unsafe`), this only works if `BerReader` is a `repr(transparent)` newtype for the inner type, and thus it can't carry any state, which I presume will be needed to track at least the length of the current indefinite segment.

And if the goal is to support arbitrary nested indefinite lengths, it might need to keep track of a stack of those.

I guess one important question is: do we need to support indefinite lengths for just terminal values, or do we need to support arbitrary nested indefinite length fields?

**Comment by Firstyear** [comment]
To my knowledge I can't think of any spec/proto that requires these to be nested, but I'm only one person :) I'd be okay without the arbitrary nesting.

**Comment by bkstein** [comment]
> I guess one important question is: do we need to support indefinite lengths for just terminal values, or do we need to support arbitrary nested indefinite length fields?

@tarcieri I'd love to say "no", but I'm pretty sure, that we need to support arbitrary nested indefinite lengths for PKCS#7/CMS messages.

A SCEP PKCS#7(CMS) message has the following structure:

```
pkiMessage {
  contentType = signedData { pkcs-7 2 },
  content {
    CMSVersion,
    digestAlgorithms,
    encapsulatedContentInfo {
      eContentType = data { pkcs-7 1 },
      eContent {           -- pkcsPKIEnvelope, optional
        contentType = envelopedData { pkcs-7 3 },
        content {
          recipientInfo,
          encryptedContentInfo {
            contentType = data { pkcs-7 1 },
            contentEncrAlgorithm,
            encryptedContent {
              messageData  -- Typically PKCS #10 request
            }
          }
        }
      }
    },
    certificates,          -- Optional
    crls,                  -- Optional
    signerInfo {
      CMSversion,
      SignerIdentifier,
      DigestAlgorithmIdentifier,
      signedAttrs {
        transactionID,
        messageType,
        pkiStatus,
        failInfo,          -- Optional
        senderNonce / recipientNonce,
      },
      SignatureAlgorithmIdentifier,
      SignatureValue,
      UnsignedAttributes -- Optional
    }
  }
}
```

Here, we have content containing encapsulated content containing encrypted content. And according to [RFC 5652](https://datatracker.ietf.org/doc/html/rfc5652#section-1) the encoding is BER. I remember running into problems with these nested structures when I tried to parse a SCEP message from EJBCA.
Our current solution for parsing BER is to replace all indefinite lengths with definite lenghts before we hand it over to RustCrypto's `der::Decode::from_der()`.

**Comment by dwhjames** [comment]
Here are 4 messages from AWS KMS.

1. [kms_ciphertext_for_recipient_1.ber.base64.txt](https://github.com/user-attachments/files/20402214/kms_ciphertext_for_recipient_1.ber.base64.txt)
1. [kms_ciphertext_for_recipient_2.ber.base64.txt](https://github.com/user-attachments/files/20402215/kms_ciphertext_for_recipient_2.ber.base64.txt)
1. [kms_ciphertext_for_recipient_3.ber.base64.txt](https://github.com/user-attachments/files/20402216/kms_ciphertext_for_recipient_3.ber.base64.txt)
1. [kms_ciphertext_for_recipient_4.ber.base64.txt](https://github.com/user-attachments/files/20402213/kms_ciphertext_for_recipient_4.ber.base64.txt)

---

A presentation of the first

[https://lapo.it/asn1js/#MIAGCSq…](https://lapo.it/asn1js/#MIAGCSqGSIb3DQEHA6CAMIACAQIxggFrMIIBZwIBAoAgcQ0zX-yOdGR8CYwov05XCbHVlfydT4HojTIC6-f7aZcwPAYJKoZIhvcNAQEHMC-gDzANBglghkgBZQMEAgEFAKEcMBoGCSqGSIb3DQEBCDANBglghkgBZQMEAgEFAASCAQApz_6xMi9PgrgYx6rvL-IP5U3G3b914l3A6Z3dxV72IoZN7bi5ILgBHJND5x1WzLCKuiLcNboY3v776NaZcCcXpD_Gtn-0X7BECFTdDkO58xKEP-24A9F8YUp2Aa6WDPR6Sf3-KAmaSy9aIiwgqLjT_gQTjYBP9AfRUtpiCZCxLTbARkTt0sArPiRef9-ltIxXKptWfbT8pOacD8orrQVZ433UajAtGOOTiAVAt8PcCBNLpCEisamJoSyPhW5peAkuulELN9bCr-2cvl62-cUsZOoPElV8kIc9b2eCnf15MD38oyX-j99xwCfZB4bpSo9cAuTftyD0kH3tva3ZIhEDMIAGCSqGSIb3DQEHATAdBglghkgBZQMEASoEEDc0PfFHDfYl7rb0v9LxrMOggAQQzHSt9l2XPItyzVHhuSfw8AAAAAAAAAAAAAA)

<details>
<summary>output…</summary>

```
ContentInfo SEQUENCE @0-459 (constructed): (2 elem)
  contentType ContentType OBJECT_IDENTIFIER @2+9: 1.2.840.113549.1.7.3|envelopedData|PKCS #7
  content [0] @13-444 (constructed): (1 elem)
    EnvelopedData SEQUENCE @15-440 (constructed): (3 elem)
      version CMSVersion INTEGER @17+1: 2
      recipientInfos RecipientInfos SET @20+363 (constructed): (1 elem)
        ktri RecipientInfo SEQUENCE @24+359 (constructed): (4 elem)
          version CMSVersion INTEGER @28+1: 2
          subjectKeyIdentifier RecipientIdentifier [0] @31+32: (32 byte)|710D335FEC8E74647C098C28BF4E5709B1D595FC9D4F81E88D3202EBE7FB6997
          keyEncryptionAlgorithm KeyEncryptionAlgorithmIdentifier SEQUENCE @65+60 (constructed): (2 elem)
            algorithm OBJECT_IDENTIFIER @67+9: 1.2.840.113549.1.1.7|rsaOAEP|PKCS #1
            parameters ANY SEQUENCE @78+47 (constructed): (2 elem)
              [0] @80+15 (constructed): (1 elem)
                SEQUENCE @82+13 (constructed): (2 elem)
                  OBJECT_IDENTIFIER @84+9: 2.16.840.1.101.3.4.2.1|sha-256|NIST Algorithm
                  NULL @95+0
              [1] @97+28 (constructed): (1 elem)
                SEQUENCE @99+26 (constructed): (2 elem)
                  OBJECT_IDENTIFIER @101+9: 1.2.840.113549.1.1.8|pkcs1-MGF|PKCS #1
                  SEQUENCE @112+13 (constructed): (2 elem)
                    OBJECT_IDENTIFIER @114+9: 2.16.840.1.101.3.4.2.1|sha-256|NIST Algorithm
                    NULL @125+0
          encryptedKey EncryptedKey OCTET_STRING @127+256: (256 byte)|29CFFEB1322F4F82B818C7AAEF2FE20FE54DC6DDBF75E25DC0E99DDDC55EF622864DEDB8B920B8011C9343E71D56CCB08ABA22DC35BA18DEFEFBE8D699702717A43FC6B67FB45FB0440854DD0E43B9F312843FEDB803D17C614A7601AE960CF47A49FDFE28099A4B2F5A222C20A8B8D3FE04138D804FF407D152DA620990B12D36C04644EDD2C02B3E245E7FDFA5B48C572A9B567DB4FCA4E69C0FCA2BAD0559E37DD46A302D18E393880540B7C3DC08134BA42122B1A989A12C8F856E6978092EBA510B37D6C2AFED9CBE5EB6F9C52C64EA0F12557C90873D6F67829DFD79303DFCA325FE8FDF71C027D90786E94A8F5C02E4DFB720F4907DEDBDADD9221103
      encryptedContentInfo EncryptedContentInfo SEQUENCE @387-66 (constructed): (3 elem)
        contentType ContentType OBJECT_IDENTIFIER @389+9: 1.2.840.113549.1.7.1|data|PKCS #7
        contentEncryptionAlgorithm ContentEncryptionAlgorithmIdentifier SEQUENCE @400+29 (constructed): (2 elem)
          algorithm OBJECT_IDENTIFIER @402+9: 2.16.840.1.101.3.4.1.42|aes256-CBC|NIST Algorithm
          parameters ANY OCTET_STRING @413+16: (16 byte)|37343DF1470DF625EEB6F4BFD2F1ACC3
        EncryptedContent [?] [0] @431-20 (constructed): (1 elem)
          OCTET_STRING @433+16: (16 byte)|CC74ADF65D973C8B72CD51E1B927F0F0
```

</details>

---

@tarcieri 

> And if the goal is to support arbitrary nested indefinite lengths, it might need to keep track of a stack of those.
>
> I guess one important question is: do we need to support indefinite lengths for just terminal values, or do we need to support arbitrary nested indefinite length fields?

I _think_ this is what you are asking, but the above terminates with a sequence of 5 EoC (2-byte 0x0000). For CMS `EnvelopedData`, the `EncryptedContent` is the value that motivates the use of BER and also happens to appear ‘last’ within the structure.

**Comment by dwhjames** [comment]
From my comment above
- https://github.com/RustCrypto/formats/issues/779#issuecomment-2646083625

> I spent some time searching for additive-only changes, and even with, PR https://github.com/RustCrypto/formats/pull/1321, I could not see a path.
>
> My impression from the codebase was that there would need to be breaking changes, and one possible path might look like,
>
> - using `IndefiniteLength` rather than `Length` within `Header`,
> - passing `Header` rather than `Length`, to function `read_nested` of trait `Reader`,
>
> so that the existing API and usage doesn’t change dramatically, but so that nested readers can handle the constructed, indefinite-length method.

I wrote my own `Header`

```rust
struct BerHeader {
    tag: Tag,
    length: IndefiniteLength,
}
```

and wrapper

```rust
struct BerSliceReader<'a> {
    inner: SliceReader<'a>,
}
```

and to work around the same observation as above:
- https://github.com/RustCrypto/formats/issues/779#issuecomment-2855695913

> Unfortunately `Reader::read_nested` uses an `FnOnce` that accepts `&mut Self` as an argument. 

I had to re-implement sections of
- https://docs.rs/cms/0.2.3/cms/enveloped_data/index.html

in terms of my `BerSliceReader`, to deal with nesting.

---

> Our current solution for parsing BER is to replace all indefinite lengths with definite lenghts before we hand it over to RustCrypto's der::Decode::from_der().

@bkstein I also attempted that

- https://github.com/RustCrypto/formats/pull/1652

but I’m closing this, because it’s not a general solution. Doing a schema-oblivious transcoding doesn’t work generally, due to implicit tags.

[RFC 5652 § 6.1. EnvelopedData Type](https://www.rfc-editor.org/rfc/rfc5652#section-6.1)
```
      EncryptedContentInfo ::= SEQUENCE {
        contentType ContentType,
        contentEncryptionAlgorithm ContentEncryptionAlgorithmIdentifier,
        encryptedContent [0] IMPLICIT EncryptedContent OPTIONAL }

      EncryptedContent ::= OCTET STRING
```

here a slice from a BER encoding from AWS KMS corresponding to `encryptedContent [0] IMPLICIT EncryptedContent OPTIONAL`

```
A0 80 04 10 CC 74 AD F6 5D 97 3C 8B 72 CD 51 E1 B9 27 F0 F0 00 00

```

the implicit tagging is hiding the fact that this is a constructed string of octet string, with 1 chunk of octet string (but can be many chunks)

in explicit tagging this would have been

```
A0 80
  14 0C
    04 10 CC 74 AD F6 5D 97 3C 8B 72 CD 51 E1 B9 27 F0 F0
00 00
```
(whitespace for clarity)

and in DER
```
A0 10 CC 74 AD F6 5D 97 3C 8B 72 CD 51 E1 B9 27 F0 F0
```
(only primitive strings allowed)
```

**Comment by tarcieri** [comment]
Yeah, thanks to `IMPLICIT` you really need a "schema aware" decoder.

I do intend to try the `BerReader` approach again, borrowing some inspiration from #1820 and adding something like `Reader::read_field` which functions like `Reader::read_nested` but has access to the whole `Header`.

**Comment by tarcieri** [comment]
With #1897 and #1899 in place, it seems like basic support for indefinite length decoding should be in place, and where there are gaps (e.g. support for constructed strings other than `OCTET STRING`) it should now be somewhat easy to fill them.

#1898 includes expanded tests against the `cms` crate which should inform usage for this particular case.

Please try it out and let me know if it's acceptable. I'm going to leave this issue open for now until I get some confirmation that things appear to be working.

**Comment by dwhjames** [comment]
I see why
- https://github.com/RustCrypto/formats/pull/1898
  - https://github.com/RustCrypto/formats/blob/master/cms/tests/examples/cms_ber.bin

Doesn’t cover the implicit case…

https://www.rfc-editor.org/rfc/rfc5652#section-5.2

```
      EncapsulatedContentInfo ::= SEQUENCE {
        eContentType ContentType,
        eContent [0] EXPLICIT OCTET STRING OPTIONAL }
```

---

Compared with

https://www.rfc-editor.org/rfc/rfc5652#section-6.1

```
      EncryptedContentInfo ::= SEQUENCE {
        contentType ContentType,
        contentEncryptionAlgorithm ContentEncryptionAlgorithmIdentifier,
        encryptedContent [0] IMPLICIT EncryptedContent OPTIONAL }

      EncryptedContent ::= OCTET STRING
```

**Comment by tarcieri** [comment]
I've thought about preserving the original `constructed` flag verbatim by adding it to `Header`

**Comment by dwhjames** [comment]
RE: my other comments with tests above,

1. constructed, definite-length octet strings; should either work correctly or return an appropriate error; I haven’t encountered these in the wild, and these don’t seem well motivated (if the length is known, then there’s a reasonable expectation to just use a primitive, definite-length octet string, and hard to see an incentive to chunk unless some internal ‘rope’ structure is being reflected into the encoding)
2. recursive constructed, indefinite-length octet strings; currently error (is it specific enough?); I haven’t encountered these in the wild, and the motivation / incentive to encode octet strings these way seems difficult to imagine (like above, maybe internal implementation details leaking into the encoding strategy).
